octree_creation_app/driver.py

```diff
@@ -72,50 +72,82 @@ def octree_from_params(params: OctreeParams):
-                    refinement_object,
-                    levels,
-                    params.diagonal_balance,
-                    max_distance=getattr(params, value["distance"]),
-                )
+        print("Finalizing . . .")
+        mesh.finalize()
+        octree = treemesh_2_octree(params.geoh5, mesh, name=params.ga_group_name)
+        print("Done.")
+        return octree
 
-            elif isinstance(refinement_object, Curve):
-                mesh = OctreeDriver.refine_tree_from_curve(
-                    mesh, refinement_object, levels, params.diagonal_balance
-                )
+    @staticmethod
+    def refine_by_object_type(
+        mesh: TreeMesh,
+        refinement_object: ObjectBase,
+        levels: list[int],
+        horizon: bool,
+        distance: float | None,
+        diagonal_balance: bool,
+    ) -> TreeMesh:
+        """Refine Treemesh as a based on object type."""
+        if horizon:
+            mesh = OctreeDriver.refine_tree_from_surface(
+                mesh,
+                refinement_object,
+                levels,
+                diagonal_balance,
+                max_distance=np.inf if distance is None else distance,
+            )
 
-            elif isinstance(refinement_object, Surface):
-                mesh = OctreeDriver.refine_tree_from_triangulation(
-                    mesh, refinement_object, levels, params.diagonal_balance
-                )
+        elif isinstance(refinement_object, Curve):
+            mesh = OctreeDriver.refine_tree_from_curve(
+                mesh, refinement_object, levels, diagonal_balance
+            )
 
-            elif isinstance(refinement_object, Points):
-                mesh = OctreeDriver.refine_tree_from_points(
-                    mesh,
-                    refinement_object,
-                    levels,
-                    diagonal_balance=params.diagonal_balance,
-                )
+        elif isinstance(refinement_object, Surface):
+            mesh = OctreeDriver.refine_tree_from_triangulation(
+                mesh, refinement_object, levels, diagonal_balance
+            )
 
-            else:
-                raise NotImplementedError(
-                    f"Refinement for object {type(refinement_object)} is not implemented."
-                )
+        elif isinstance(refinement_object, Points):
+            mesh = OctreeDriver.refine_tree_from_points(
+                mesh,
+                refinement_object,
+                levels,
+                diagonal_balance=diagonal_balance,
+            )
 
-        print("Finalizing . . .")
-        mesh.finalize()
-        octree = treemesh_2_octree(params.geoh5, mesh, name=params.ga_group_name)
+        else:
+            raise NotImplementedError(
+                f"Refinement for object {type(refinement_object)} is not implemented."
+            )
 
-        return octree
+        if hasattr(refinement_object, "complement"):
+            mesh = OctreeDriver.refine_by_object_type(
```

Comment:
Unfortunately, this sends it into an infinite loop. You will need an argument to only execute the complement once.

Reply:
using a loop now instead